File: connectors/inmem/sync/src/main/java/com/coditory/sherlock/inmem/InMemorySherlock.java
package com.coditory.sherlock.inmem;

import com.coditory.sherlock.Sherlock;
import com.coditory.sherlock.SherlockDefaults;
import com.coditory.sherlock.SherlockWithConnectorBuilder;
import org.jetbrains.annotations.NotNull;

import java.time.Clock;

import static com.coditory.sherlock.Preconditions.expectNonNull;
import static com.coditory.sherlock.SherlockDefaults.DEFAULT_CLOCK;

/**
 * Builds {@link Sherlock} that that stores locks in memory.
 * <p>
 * Designed for testing purposes only.
 */
public final class InMemorySherlock extends
    SherlockWithConnectorBuilder<InMemorySherlock> {
    private Clock clock = DEFAULT_CLOCK;
    private InMemoryDistributedLockStorage storage = new InMemoryDistributedLockStorage();

    /**
     * @return new instance of the builder
     */
    @NotNull
    public static InMemorySherlock builder() {
        return new InMemorySherlock();
    }

    /**
     * @return new instance iof in memory sherlock with default configuration
     */
    @NotNull
    public static Sherlock create() {
        return InMemorySherlock.builder().build();
    }

    private InMemorySherlock() {
        // deliberately empty
    }

    /**
     * @param clock time provider used in locking mechanism. Default: {@link
     *              SherlockDefaults#DEFAULT_CLOCK}
     * @return the instance
     */
    @NotNull
    public InMemorySherlock withClock(@NotNull Clock clock) {
        this.clock = expectNonNull(clock, "clock");
        return this;
    }

    /**
     * Use shared stage for all instances of {@link Sherlock}.
     *
     * @return the instance
     */
    @NotNull
    public InMemorySherlock withSharedStorage() {
        this.storage = InMemoryDistributedLockStorage.singleton();
        return this;
    }

    /**
     * @return sherlock instance
     * @throws IllegalArgumentException when some required values are missing
     */
    @Override
    @NotNull
    public Sherlock build() {
        InMemoryDistributedLockConnector connector = new InMemoryDistributedLockConnector(
            clock, storage);
        return super.build(connector);
    }
}

File: connectors/inmem/coroutines/src/main/kotlin/com/coditory/sherlock/inmem/coroutines/InMemorySherlock.kt
package com.coditory.sherlock.inmem.coroutines

import com.coditory.sherlock.Preconditions.expectNonNull
import com.coditory.sherlock.SherlockDefaults
import com.coditory.sherlock.coroutines.Sherlock
import com.coditory.sherlock.coroutines.SherlockWithConnectorBuilder
import com.coditory.sherlock.inmem.InMemoryDistributedLockStorage
import java.time.Clock

/**
 * Builds [Sherlock] that that stores locks in memory.
 *
 * Designed for testing purposes only.
 */
class InMemorySherlock private constructor() : SherlockWithConnectorBuilder<InMemorySherlock>() {
    private var clock = SherlockDefaults.DEFAULT_CLOCK
    private var storage = InMemoryDistributedLockStorage()

    /**
     * @param clock time provider used in locking mechanism. Default: [SherlockDefaults.DEFAULT_CLOCK]
     * @return the instance
     */
    fun withClock(clock: Clock): InMemorySherlock {
        this.clock = expectNonNull(clock, "clock")
        return this
    }

    /**
     * Use shared stage for all instances of [Sherlock].
     *
     * @return the instance
     */
    fun withSharedStorage(): InMemorySherlock {
        storage = InMemoryDistributedLockStorage.singleton()
        return this
    }

    /**
     * @return sherlock instance
     * @throws IllegalArgumentException when some required values are missing
     */
    override fun build(): Sherlock {
        val connector =
            KtInMemoryDistributedLockConnector(
                clock,
                storage,
            )
        return super.build(connector)
    }

    companion object {
        /**
         * @return new instance of the builder
         */
        @JvmStatic
        fun builder(): InMemorySherlock {
            return InMemorySherlock()
        }

        /**
         * @return new instance iof in memory sherlock with default configuration
         */
        @JvmStatic
        fun create(): Sherlock {
            return builder().build()
        }
    }
}
